sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala

```diff
@@ -283,46 +344,61 @@ case class GpuProjectAstExec(
-  // Returns a list of boolean values the same size as the first
-  // sequence, where the values are true if the corresponding attribute can be skipped.
-  // The indices of these attributes correspond to the columns that were bound via
-  // GpuBindReferences.bindGpuReference.
-  private def getColumnSkips(inputTiers: Seq[AttributeSeq]): Array[Boolean] = inputTiers match {
-    case Nil => Array.emptyBooleanArray
-    case curTier :: Nil =>
-      // For the last tier, fill with all true (skip) values, because the output
-      // of the last tier does not include any input columns.
-      Array.fill(curTier.attrs.size)(true)
-    case curTier :: tail =>
-      val curAttrs = curTier.attrs
-      val nextAttrs = tail.head.attrs
-      // This is equivalent to:
-      // skipList = curAttrs.map(a => if (nextAttrs.contains(a)) false else true)
-      // but this should be faster
-      val skipList = new Array[Boolean](curAttrs.size)
-      var curIdx = 0
-      var nextIdx = 0
-      while (curIdx < curAttrs.size) {
-        if (nextAttrs(nextIdx) == curAttrs(curIdx)) {
-          skipList(curIdx) = false
-          nextIdx += 1
-        } else {
-          skipList(curIdx) = true
+ case class GpuTieredProject(exprTiers: Seq[Seq[GpuExpression]]) extends Arm {
+
+  /**
+   * Is everything deterministic. This can help with reliability in the common case.
+   */
+  private lazy val areAllDeterministic = !exprTiers.exists { tier =>
+    tier.exists { expr =>
+      !expr.deterministic
+    }
+  }
+
+  def projectAndCloseWithRetrySingleBatch(sb: SpillableColumnarBatch,
+      spillCallback: SpillCallback): ColumnarBatch = {
+    if (areAllDeterministic) {
+      // If all of the expressions are deterministic we can just run everything are retry it
+      // at the top level. If some things are non-deterministic we need to split them up and
+      // do the processing in a way that makes it so retries are more likely to succeed.
+      RmmRapidsRetryIterator.withRetryNoSplit(sb) { sb =>
+        withResource(sb.getColumnarBatch()) { cb =>
+          project(cb)
         }
-        curIdx += 1
       }
-      skipList
+    } else {
+      @tailrec
+      def recurse(boundExprs: Seq[Seq[GpuExpression]],
+          sb: SpillableColumnarBatch,
+          isFirst: Boolean): SpillableColumnarBatch = boundExprs match {
+        case Nil =>
+          if (isFirst) {
+            new JustRowsColumnarBatch(sb.numRows(), spillCallback)
+          } else {
+            sb
+          }
+        case exprSet :: tail =>
+          val projectSb = withResource(new NvtxRange("project tier", NvtxColor.ORANGE)) { _ =>
+            val projectResult = GpuProjectExec.projectAndCloseWithRetrySingleBatch(sb,
+              exprSet, spillCallback)
+            SpillableColumnarBatch(projectResult, SpillPriorities.ACTIVE_ON_DECK_PRIORITY,
+              spillCallback)
+          }
+          // Close intermediate batches
+          if (!isFirst) sb.close()
```

Comment:
I think this is double closing `sb`, but it is ok because RapidsBufferHandle protects against that, that said we shouldn't need this line.